### PR TITLE
Correct forward AD for linalg.eig and add checks

### PR DESCRIPTION
**Pull request by lezcano**
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #70528
* __->__ #70527

This PR adds checks for the backward of `linalg.eig`, similar to those
deduced in https://github.com/pytorch/pytorch/pull/70253

It also modifies the function so that it does not save the input matrix,
as it's not necessary.

It also corrects the forward AD formula for it to be correct. Now all
the tests pass for `linalg.eig` and `linalg.eigvals`.

It also updates the docs to reflect better what's going on here.

cc @jianyuh @nikitaved @pearu @mruberry @walterddr @IvanYashchuk @xwang233 @Lezcano

Differential Revision: [D33530148](https://our.internmc.facebook.com/intern/diff/D33530148)

### CODE REVIEW

File: torch/csrc/autograd/FunctionsManual.cpp [hunk 1]
```diff
@@ -2563,65 +2564,84 @@ Tensor linalg_eig_backward(const std::vector<torch::autograd::Variable> &grads,
...
+  }();
+  const auto VhgV = at::matmul(V.mH(), gV);
+  const auto diag_VhgV = VhgV.diagonal(0, -2, -1);
 
-    const auto VhgV = at::matmul(V.mH(), gV);
-
-    const auto diag_re_VhgV = at::real(VhgV).diagonal(0, -2, -1);
-    auto result = VhgV - at::matmul(V.mH(), V * diag_re_VhgV.unsqueeze(-2));
+  // Check invariance of the loss function wrt the transformation V -> V e^{i\phi}
+  if (is_complex) {
```
Comment (IvanYashchuk) [inline torch/csrc/autograd/FunctionsManual.cpp:2594]: Eigenvectors can be complex also for real input matrices.

**Review by IvanYashchuk — COMMENTED** [review]
Invariance of the loss function is checked only for the complex matrix case, but for real inputs eigenvectors are complex.

**Comment by Lezcano** [comment]
@albanD this stack is ready to get a last review / be landed. There are reds in the CI, but they are unrelated to these PRs.

**Review by albanD — APPROVED** [review]
Ok!

**Review by IvanYashchuk — APPROVED** [review]
